### PR TITLE
Article options dropdown

### PR DESCRIPTION
**Pull request by charliew-rsi**
This PR is based off submit-article-middle-ware. It adds a drop-down for snippets. When the "Stack Overflow Post" is selected as the type of article, the form on the "Submit Post" page replaces the Summary field with the Snippets field. The Snippets field will store pasted text as a plain text object and formats the text before displaying it in a tab with similar funcitionality to the Tags tabs, albeit different styling. Additional functionality was added to reset all the form inputs after submitting an article.

### CODE REVIEW

File: public/styles/Editor.css [hunk 1]
```diff
@@ -0,0 +1,26 @@
+.snippet-container {    
+    font-size: 0.8rem;
+    padding-top: 0.2rem;
+    padding-bottom: 0.2rem;
+    white-space: nowrap;
+    margin-bottom: 0.2rem;
+    background-color: #f3f3f3;
+    color: #818a91 !important;
```
Comment (wesdollar) [inline public/styles/Editor.css]: Avoid using `!important`. Opt for higher specificity instead.

File: public/styles/Editor.css [hunk 2]
```diff
@@ -0,0 +1,26 @@
+.snippet-container {
```
Comment (wesdollar) [inline public/styles/Editor.css]: It would be much better to opt for shorthand syntax for padding and margin.

File: src/components/Editor.js [hunk 3]
```diff
@@ -1,97 +1,105 @@
...
 
-  componentWillMount() {
-    if (this.props.match.params.slug) {
-      return this.props.onLoad(agent.Articles.get(this.props.match.params.slug));
+  watchForEnter(event) {
+    event.preventDefault();
+    if (event.keyCode === 13) {
+      switch(event.target.name) {
+        case "tagInput" :
+          this.setState(prevState => ({tags: [...prevState.tags, prevState.tagInput], tagInput: ""}));
```
Comment (wesdollar) [inline src/components/Editor.js]: Per conversation, double-check the functionality here. It's a bit odd to me. Using React dev tools, confirm your state is correct as you intend it to be after it is updated.

File: src/components/Editor.js [hunk 4]
```diff
@@ -1,97 +1,105 @@
...
-      return this.props.onLoad(agent.Articles.get(this.props.match.params.slug));
+  watchForEnter(event) {
+    event.preventDefault();
+    if (event.keyCode === 13) {
+      switch(event.target.name) {
+        case "tagInput" :
+          this.setState(prevState => ({tags: [...prevState.tags, prevState.tagInput], tagInput: ""}));          
+        break;
+        case "snippetInput" :
+          this.setState(prevState => ({snippets: [...prevState.snippets, prevState.snippetInput], snippetInput: ""}));
```
Comment (wesdollar) [inline src/components/Editor.js]: Per conversation, double-check the functionality here. It's a bit odd to me. Using React dev tools, confirm your state is correct as you intend it to be after it is updated.

File: src/components/Editor.js [hunk 5]
```diff
@@ -1,97 +1,105 @@
...
+      switch(event.target.name) {
+        case "tagInput" :
+          this.setState(prevState => ({tags: [...prevState.tags, prevState.tagInput], tagInput: ""}));          
+        break;
+        case "snippetInput" :
+          this.setState(prevState => ({snippets: [...prevState.snippets, prevState.snippetInput], snippetInput: ""}));          
+        break;
+
+        default: 
+        break;
```
Comment (wesdollar) [inline src/components/Editor.js]: `default` doesn't need a break. [MDN switch](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch)

File: src/components/Editor.js [hunk 6]
```diff
@@ -1,97 +1,105 @@
...
-  componentWillMount() {
-    if (this.props.match.params.slug) {
-      return this.props.onLoad(agent.Articles.get(this.props.match.params.slug));
+  watchForEnter(event) {
+    event.preventDefault();
+    if (event.keyCode === 13) {
+      switch(event.target.name) {
+        case "tagInput" :
+          this.setState(prevState => ({tags: [...prevState.tags, prevState.tagInput], tagInput: ""}));          
+        break;
```
Comment (wesdollar) [inline src/components/Editor.js]: `break` statements should be indented

File: src/components/Editor.js [hunk 7]
```diff
@@ -1,97 +1,105 @@
...
+    arr.splice(index, 1);
+    this.setState({[event.target.name] : arr});
+  };
+
+  isDisabled() {
+    const {url, title, summary, tags, type, snippets} = this.state;
+    if (type !== "Stack Overflow Post" && (url.length === 0 || title.length === 0 || summary.length === 0 || tags.length === 0)) {
+      return true;
+    }
+    else if (type === "Stack Overflow Post" && (url.length === 0 || title.length === 0 || snippets.length === 0 || tags.length === 0)) {
```
Comment (wesdollar) [inline src/components/Editor.js]: Based on conversation, clean this stuff up.

File: src/components/Editor.js [hunk 8]
```diff
@@ -103,63 +111,121 @@ class Editor extends React.Component {
...
-                    </div>
+                        )}
+                    </div>                        
                   </fieldset>
-
                   <button
                     className="btn btn-lg pull-xs-right btn-primary"
                     type="button"
-                    disabled={this.props.inProgress}
+                    disabled={this.isDisabled()}
```
Comment (wesdollar) [inline src/components/Editor.js]: I don't believe the `()` is necessary here. Including the `()` causes the function to fire on render, which I don't believe is your intended behavior.

File: src/components/services/article.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+const requests = {
+    post: (url, body) => superagent.post(`${API_ROOT}${url}`, body).then(responseBody)
+};
+
+const SubmitArticle = article => {
+    requests.post("/article/submit", article);    
+};
+
+export default SubmitArticle;
```
Comment (wesdollar) [inline src/components/services/article.js]: New line at end of file

File: src/data/articles.json [hunk 10]
```diff
@@ -0,0 +1,3 @@
+[{
+    
+}]
```
Comment (wesdollar) [inline src/data/articles.json]: new line at end of file

File: src/components/Editor.js (hunk 8 above)
Comment (charliew-rsi) [reply src/components/Editor.js]: Tested functionality without the `()`, it is required for the intended behavior. In the original program, inProgress is available to the store and is evaluated before the render. I opted to do more validation locally in the component on the front-end ergo we do want it to re-evaluate when re-rendered.

File: public/styles/Editor.css (hunk 1 above)
Comment (charliew-rsi) [reply public/styles/Editor.css]: `!important` ended up being superfluous

File: src/components/Editor.js (hunk 3 above)
Comment (charliew-rsi) [reply src/components/Editor.js]: Correct functionality confirmed.

File: public/styles/Editor.css (hunk 2 above)
Comment (charliew-rsi) [reply public/styles/Editor.css]: resolved

File: src/components/Editor.js (hunk 4 above)
Comment (charliew-rsi) [reply src/components/Editor.js]: functionality confirmed

File: src/data/articles.json (hunk 10 above)
Comment (charliew-rsi) [reply src/data/articles.json]: deleting file, will exist on back-end service

**Pull request closed without merging**